File: api/src/test/java/net/blueshell/api/factory/model/RecoveryTokenFactory.java
package net.blueshell.api.factory.model;

import net.blueshell.api.testutil.ModelTestUtils;
import com.github.javafaker.Faker;
import lombok.RequiredArgsConstructor;
import net.blueshell.api.common.enums.ResetType;
import net.blueshell.api.model.RecoveryToken;
import net.blueshell.api.model.User;
import org.springframework.stereotype.Component;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Factory for RecoveryToken model test instances.
 */
@Component
@RequiredArgsConstructor
public class RecoveryTokenFactory {

    private static final AtomicLong counter = new AtomicLong(1000);
    private final Faker faker;
    private final UserFactory userFactory;

    public RecoveryToken createBasic() {
        RecoveryToken rt = new RecoveryToken();
        User u = userFactory.createBasic();
        rt.setUser(u);
        rt.setType(faker.options().option(ResetType.class));
        rt.setSelector(faker.crypto().sha256().substring(0, 32));
        rt.setVerifierHash(faker.crypto().sha256());
        rt.setExpiresAt(Instant.now().plus(2, ChronoUnit.HOURS));
        rt.setConsumedAt(null);
        return rt;
    }

    public RecoveryToken createFull() {
        return createBasic();
    }

    public RecoveryToken createWithCustomizations(java.util.function.Consumer<RecoveryToken> customizer) {
        RecoveryToken rt = createFull();
        customizer.accept(rt);
        return rt;
    }

    private Long generateId() {
        return counter.incrementAndGet();
    }
}

File: api/src/test/kotlin/net/blueshell/api/factory/model/RecoveryTokenFactory.kt
package net.blueshell.api.factory.model

import com.github.javafaker.Faker
import net.blueshell.api.shared.enums.ResetType
import net.blueshell.api.auth.model.RecoveryToken
import org.springframework.stereotype.Component
import java.time.Instant
import java.time.temporal.ChronoUnit
import java.util.concurrent.atomic.AtomicLong
import java.util.function.Consumer

/**
 * Factory for RecoveryToken model test instances.
 */
@Component
class RecoveryTokenFactory(
    private val faker: Faker,
    private val userFactory: UserFactory
) {

    fun createBasic(): RecoveryToken {
        val token = RecoveryToken()
        val user = userFactory.createBasic()
        token.user = user
        token.type = faker.options().option(ResetType::class.java)
        token.selector = faker.crypto().sha256().substring(0, 32)
        token.verifierHash = faker.crypto().sha256()
        token.expiresAt = Instant.now().plus(2, ChronoUnit.HOURS)
        token.consumedAt = null
        return token
    }

    fun createFull(): RecoveryToken = createBasic()

    fun createWithCustomizations(customizer: Consumer<RecoveryToken>): RecoveryToken {
        val token = createFull()
        customizer.accept(token)
        return token
    }

    private fun generateId(): Long = COUNTER.incrementAndGet()

    private companion object {
        val COUNTER = AtomicLong(1000)
    }
}
